### ISSUE TITLE
Changes in ChannelArguments do not reflect when done during execution

### ISSUE THREAD

**Issue by ParmeshMathur** [grpc/grpc#33174]
### What version of gRPC and what language are you using?
`gRPC v 1.52.1 / C++`

### What operating system (Linux, Windows,...) and version?
`Windows 10 Pro, version 22H2`

### What runtime / compiler are you using (e.g. python version or version of gcc)
`Visual Studio Professional 2019`

### What did you do?
1. A basic program to send an RPC request to a server from the client - directly and via a proxy.
2. The Proxy has an unreachable address and any request made to it (or via it) will fail.
3. Using `ChannelArguments` to `CreateCustomChannel()` with a parameter called "proxy" in the ChannelArguments that can be enabled or disabled (set to 1 or 0).
4. Once the channel has been initialised with the proxy enabled mode using `SetInt("proxy", 1)`, an RPC request is sent. Then, after using `SetInt("proxy", 0)` to avoid using the proxy, another RPC request is sent (directly to the server).

### What did you expect to see?
The RPC request must fail the first time (on requesting via the _invalid_ proxy) and succeed the second time (on direct access).

### What did you see instead?
The RPC request fails both times, indicating that the proxy 'mode' in the channel arguments has not been changed. This remains the same unless a new object of type `ChannelArguments` is created with the another mode.

Make sure you include information that can help us debug (full error message, exception listing, stack trace, logs).

See [TROUBLESHOOTING.md](https://github.com/grpc/grpc/blob/master/TROUBLESHOOTING.md) for how to diagnose problems better.

### Anything else we should know about your project / environment?
Nothing

**Comment by yijiem** [comment]
Sorry for the late response. Are you using the right channel args to set the http proxy? You can use the macros specified in: https://github.com/grpc/grpc/blob/master/include/grpc/impl/grpc_types.h#L440-L444

Also, can you attach error logs? What errors are you seeing? It would be easier for us to diagnose the issue.

**Comment by ParmeshMathur** [comment]
Yes, I am using these macros, e.g. `SetInt(GRPC_ARG_ENABLE_HTTP_PROXY, 1)` and `SetInt(GRPC_ARG_ENABLE_HTTP_PROXY, 0)`
Sorry for not making that clear.

There is no error log because there aren't any 'errors' as such, just a bug/difference in interpretation.

The value held against the key (which is defined in the macro) is set to a new one. But still, the old value is showing up as the one associated with the key.

What was noticed upon digging a little into the [implementation(https://chromium.googlesource.com/external/github.com/grpc/grpc/+/chromium-deps/2016-07-19/src/cpp/common/channel_arguments.cc#:~:text=void%20ChannelArguments%3A%3ASetInt,%7D) of the `SetInt()` method, is that the function is not changing the value associated to the key that is already present, but adding a new entry in the list with the new value. And on reading the data, the value that is associated with the first occurrence of this key is what is retrieved. Hence the value that is set first is the once that is always fetched on querying the list.

**Comment by yijiem** [comment]
Is your use case one of the 2 scenarios below? If not, what is your use case?

Scenario 1 (use the same ChannelArgument and the same channel):
```
ChannelArguments args;
args.SetInt(GRPC_ARG_ENABLE_HTTP_PROXY, 1);
auto channel = grpc::CreateCustomChannel(..., args);
// use channel to send an RPC
args.SetInt(GRPC_ARG_ENABLE_HTTP_PROXY, 0);
// use the same channel again
```

Scenario 2 (use the same ChannelArgument but different channels):
```
ChannelArguments args;
args.SetInt(GRPC_ARG_ENABLE_HTTP_PROXY, 1);
auto channel1 = grpc::CreateCustomChannel(..., args);
// use channel1 to send an RPC
args.SetInt(GRPC_ARG_ENABLE_HTTP_PROXY, 0);
auto channel2 = grpc::CreateCustomChannel(..., args);
// use channel2 to send an RPC
```

I don't think we support Scenario 1. After the channel has been created, modifying the ChannelArgument should not change the behavior of the channel. If your use case is Scenario 2, then there might be something we can do to make this part of our API easier to work with (you may use a separate ChannelArgument for now).

**Comment by ParmeshMathur** [comment]
Currently, I am using the work-around (of having a separate ChannelArgument for each case).
If there is a way to provide a fix such that we can make scenario 2 work, that will certainly be helpful.

**Comment by pawbhard** [comment]
For channel args when we set multiple values for same key, expected behavior is to pick the first value. Hence, changes to the key is not being observed.

```
args.SetInt(GRPC_ARG_ENABLE_HTTP_PROXY, 1);
args.SetInt(GRPC_ARG_ENABLE_HTTP_PROXY, 0);
```

code reference : https://github.com/grpc/grpc/blob/master/src/core/lib/channel/channel_args.cc#L673
Commit reference: https://github.com/grpc/grpc/commit/09e7e7456b2b976815be98b568b0043c5d1d4e13
